modules/lily-compiler/src/main/java/io/github/tomboyo/lily/compiler/icg/OasSchemaToAst.java

```diff
@@ -270,11 +262,95 @@ class package.MyClass,
+    }
+
+    if (schema instanceof ComposedSchema c) {
+      Stream.of(c.getAllOf(), c.getAnyOf(), c.getOneOf())
+          .filter(Objects::nonNull)
+          .flatMap(List::stream)
+          .forEach(s -> properties.putAll(getProperties(s)));
+    }
+
+    return properties;
+  }
+
+  private Set<String> getMandatoryPropertyNames(Schema<?> root) {
+    var required = requiredPropertyNames(root);
+    var nonNullable = getNonNullablePropertyNames(root);
+    var mandatory = intersection(required, nonNullable);
+
+    if (root instanceof ComposedSchema c) {
+      /* If all the oneOf schema (one of which MUST validate) agree that a given property is mandatory, then it is
+      mandatory on the composed schema as well. */
+      Stream.ofNullable(c.getOneOf())
+          .flatMap(List::stream)
+          .map(this::getMandatoryPropertyNames)
+          .reduce(this::intersection)
+          .ifPresent(mandatory::addAll);
+
+      /* If a property is mandatory according to an allOf component, then it is mandatory according to the composed
+      schema as well. */
+      Stream.ofNullable(c.getAllOf())
+          .flatMap(List::stream)
+          .map(this::getMandatoryPropertyNames)
+          .forEach(mandatory::addAll);
+    }
+
+    return mandatory;
+  }
+
+  private Set<String> requiredPropertyNames(Schema<?> root) {
+    var required = new HashSet<>(requireNonNullElse(root.getRequired(), List.of()));
+
+    /* AllOf component required keywords are flattened into the composed schema. */
+    if (root instanceof ComposedSchema c) {
+      Stream.ofNullable(c.getAllOf())
+          .flatMap(List::stream)
+          .map(s -> (Schema<?>) s) // *internal screaming*
+          .map(this::requiredPropertyNames)
+          .flatMap(Set::stream)
+          .forEach(required::add);
+
+      Stream.ofNullable(c.getOneOf())
+          .flatMap(List::stream)
+          .map(s -> (Schema<?>) s) // *additional internal screaming*
+          .map(this::requiredPropertyNames)
+          .reduce(this::intersection)
+          .ifPresent(required::addAll);
+    }
+
+    return required;
+  }
+
+  private <T> HashSet<T> intersection(Collection<T> a, Collection<T> b) {
+    var result = new HashSet<>(a);
+    result.retainAll(b);
+    return result;
+  }
+
+  private Set<String> getNonNullablePropertyNames(Schema<?> root) {
+    if (root.getProperties() != null) {
+      return root.getProperties().entrySet().stream()
+          .filter(entry -> !requireNonNullElse(entry.getValue().getNullable(), false))
+          .map(Map.Entry::getKey)
+          .collect(Collectors.toSet());
+    }
+    return Set.of();
+  }
+
   /*
    * OneOf composed schemas evaluate to a sealed interface. The schema may
    * be composed of new or existing schemas; in either case, we emit
    * AddInterface modifiers so that those AST are updated to
    * implement the new sealed interface before rendering.
+   *
+   * TODO: remove me
```

Comment:
Create ticket for ADR 0002 overloaded properties using sealed interfaces and attach this to that?